### PR TITLE
Add filestream input reader

### PR DESCRIPTION
**Pull request by kvch**
## What does this PR do?

This PR adds the event reader and publisher functionality. This is mostly the refactoring of `Harvester` from `filebeat/input/log`. Two things are missing: metrics and special readers e.g. `multiline`.

## Why is it important?

New step.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~
~~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## Related issues

- Requires #21479

### CODE REVIEW

File: filebeat/input/filestream/input.go [hunk 1]
```diff
@@ -64,5 +128,240 @@ func (inp *filestream) Run(
...
+	return nil
+}
+
+func (inp *filestream) initFileOffset(file *os.File, offset int64) error {
+	if offset > 0 {
+		_, err := file.Seek(offset, os.SEEK_SET)
+		return err
+	}
+
+	_, err := file.Seek(0, os.SEEK_CUR)
```
Comment (faec) [inline filebeat/input/filestream/input.go:273]: This second `Seek` looks like a no-op. Is it testing whether the given file can support seeking in general? (If so, why do we need to know that here?) If this check is necessary please add an explanatory comment

**Review by faec — APPROVED** [review]
one minor comment

File: filebeat/input/filestream/input.go (hunk 1 above)
Comment (kvch) [reply filebeat/input/filestream/input.go:273]: TBH it is part of the original code in the `log` input. It does not make sense to have it here, but I was afraid to remove it. :D

Comment (kvch) [reply filebeat/input/filestream/input.go:274]: Added comment.

**Pull request merged**